[IMAGINATION_NNA] Add Imagination NNA support.

jiansowa opened the pull request:
First commit. Mobilenet V1 has been tested.

jiansowa committed:
NNA integration on develop branch
1, Remove a paragraph at the end of lite/CMakeLists.txt, otherwise build
fails with original developbranch on Ubuntu 18.04;
2, Add NNA support, build light library successfully;
3, Full build, library ok, but apps such as benchmark_bin failed

jiansowa committed:
1, Fix undefined reference of symbol about context;
2, Fix benchmark_bin link error that caused by missing deps;

jiansowa committed:
1, Rename dir name from nna to imagination_nna
2, Add imagination/Imagination/IMAGINATION prefix for symbol definition;
3, No build error

jiansowa committed:
1, Change bridge code to follow the new way(2.6->dev) to obtain input/output scale;
2, Add kImaginationNNA to BindTargets of
'quantized_op_attribute_inference_pass'
3, Test image classification with one imput passed

jiansowa committed:
Use IMAGINATION_NNA_SDK_ROOT in imgdnn_manager.cc

jiansowa committed:
1, revert IMGINATION_NNA_SDK_ROOT usage in imgdnn_manager.cc
2, fix output scale attribute query in op_lite.cc

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/PaddlePaddle/Paddle-Lite?pullRequest=4335) <br/>All committers have signed the CLA.

hong19860320 commented on lite/api/CMakeLists.txt:
```diff
@@ -81,7 +82,10 @@ else()
-
+        if (LITE_WITH_IMAGINATION_NNA)
+            # Need to add IMG IMAGINATION_NNA runtime libs (libhiai.so) dependency
+            #target_link_libraries(paddle_light_api_shared ${nna_builder_libs} ${nna_runtime_libs})
```
这部分注释掉也没问题吗？

jiansowa replied:
没问题。在add_kernel的时候已经建立了dependency吧。

hong19860320 commented on lite/api/CMakeLists.txt:
```diff
@@ -114,6 +118,11 @@ if(LITE_WITH_RKNPU)
 endif()
 
+if(LITE_WITH_IMAGINATION_NNA)
+    set(light_api_deps ${light_api_deps} ${nna_deps})
```
修改成imagination_nna_deps吧

jiansowa replied:
Done

hong19860320 commented on lite/backends/imagination_nna/imgdnn_manager.cc:
```diff
@@ -0,0 +1,396 @@
+
+  std::string options_str;
+  std::string ddk_root{"/home/jasonwang/imgtools/imagination_nna_sdk/"};
+  // std::string ddk_root{STR2(IMAGINATION_NNA_SDK_ROOT)};
```
去掉注释的代码

jiansowa replied:
Done

hong19860320 commented on lite/backends/imagination_nna/imgdnn_manager.cc:
```diff
@@ -0,0 +1,396 @@
+  std::string mapconfig = ddk_root + "nna-tools/config/mapconfig_q8a.json";
+  options_str += "-h " + hwconfig;
+  options_str += " -m " + mapconfig;
+  // options_str += " --dump_debug_binaries enabled";
```
去掉注释的代码

jiansowa replied:
Done.

hong19860320 commented on lite/kernels/imagination_nna/bridges/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+#lite_cc_library(subgraph_bridge_scale_op_imagination_nna SRCS scale_op.cc DEPS ${imagination_nna_subgraph_bridge_deps})
+#lite_cc_library(subgraph_bridge_softmax_op_imagination_nna SRCS softmax_op.cc DEPS ${imagination_nna_subgraph_bridge_deps})
+lite_cc_library(subgraph_bridge_pool_op_imagination_nna SRCS pool_op.cc DEPS ${imagination_nna_subgraph_bridge_deps})
+#lite_cc_library(subgraph_bridge_batch_norm_op_imagination_nna SRCS batch_norm_op.cc DEPS ${imagination_nna_subgraph_bridge_deps})
```
去掉注释掉的target

jiansowa replied:
Done.

hong19860320 commented on lite/kernels/imagination_nna/bridges/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+        subgraph_bridge_graph_imagination_nna
+        subgraph_bridge_fc_op_imagination_nna
+        subgraph_bridge_conv_op_imagination_nna
+        #subgraph_bridge_matmul_op_imagination_nna
```
去掉注释了的target

jiansowa replied:
Done.

hong19860320 commented on lite/kernels/imagination_nna/bridges/batch_norm_op.cc:
```diff
@@ -0,0 +1,92 @@
+  float epsilon = op_info->GetAttr<float>("epsilon");
+  // int mode = 1;  // bnScale, bnBias tensor dims are 1xCx1x1
+  /*
+  bool use_global_stats = !op_info->HasAttr("use_global_stats") ||
```
去掉注释的代码，如果不支持，建议增加warning

jiansowa replied:
Done.

hong19860320 commented on lite/kernels/imagination_nna/bridges/batch_norm_op.cc:
```diff
@@ -0,0 +1,92 @@
+  bn_out = builder.createScaleLayer(
+      bn_out, true, scale->mutable_data<float>(), bias->mutable_data<float>());
+
+  // PrecisionType precision = x->precision();
```
去掉注释了的代码

jiansowa replied:
Done

hong19860320 commented on lite/kernels/imagination_nna/bridges/graph.h:
```diff
@@ -0,0 +1,145 @@
+
+  ~Graph() { std::cout << "Graph deconst" << std::endl; }
+
+  // Add 1
```
如果需要，建议增加注释，而不是增加类似"Add 1/2/3"之类的无意义的注释

jiansowa replied:
Done

hong19860320 commented on lite/kernels/imagination_nna/bridges/paddle_use_bridges.h:
```diff
@@ -0,0 +1,22 @@
+USE_SUBGRAPH_BRIDGE(depthwise_conv2d, kImaginationNNA);
+USE_SUBGRAPH_BRIDGE(fc, kImaginationNNA);
+USE_SUBGRAPH_BRIDGE(pool2d, kImaginationNNA);
+// USE_SUBGRAPH_BRIDGE(softmax, kImaginationNNA);
```
去掉注释了的op bridge

jiansowa replied:
Done

hong19860320 commented on lite/kernels/imagination_nna/bridges/pool_op.cc:
```diff
@@ -0,0 +1,148 @@
+                                 ksize);
+
+  // ceil mode
+  /* bool ceil_mode =
```
如果不支持，建议增加warning

jiansowa replied:
Done

hong19860320 commented on lite/kernels/imagination_nna/bridges/utility.cc:
```diff
@@ -0,0 +1,67 @@
+  qnt->count = 0;
+  qnt->axis = 0;
+  qnt->scales.clear();
+  // qnt.scales.shrink_to_fit();
```
去掉注释了的代码

jiansowa replied:
done

hong19860320 commented on lite/kernels/imagination_nna/subgraph_compute.cc:
```diff
@@ -0,0 +1,252 @@
+  imgdnn_mgr_.getNetworkObjectInputs(
+      std::numeric_limits<unsigned int>::max(), nullptr, &num_inputs);
+  CHECK_EQ(num_inputs, device_inames_.size());
+  // origin_idims_.resize(num_inputs);
```
去掉注释的代码

jiansowa replied:
done

hong19860320 commented on lite/kernels/imagination_nna/subgraph_compute.cc:
```diff
@@ -0,0 +1,252 @@
+    auto node = graph.Get(device_inames_[i]);
+    auto type = node->type();
+    auto layout = node->layout();
+    // origin_itensors_[i] = scope_->FindMutableTensor(device_inames_[i]);
```
同上

jiansowa replied:
done

hong19860320 commented on lite/kernels/imagination_nna/subgraph_compute.cc:
```diff
@@ -0,0 +1,252 @@
+      std::numeric_limits<unsigned int>::max(), nullptr, &num_outputs);
+  CHECK_EQ(num_outputs, device_onames_.size());
+  // origin_odims_.resize(num_outputs);
+  // origin_otensors_.resize(num_outputs);
```
同上

jiansowa replied:
done

hong19860320 commented on lite/kernels/imagination_nna/subgraph_compute.cc:
```diff
@@ -0,0 +1,252 @@
+    auto type = node->type();
+    auto layout = node->layout();
+    // origin_otensors_[i] = scope_->FindMutableTensor(device_onames_[i]);
+    // CHECK(origin_otensors_[i]);
```
同上

jiansowa replied:
done

hong19860320 commented on tools/codestyle/clang_format.hook:
```diff
@@ -1,7 +1,7 @@
 set -e
 
-readonly VERSION="3.8"
+readonly VERSION="6.0.0"
```
这个不能修改，统一使用同一个clang_format版本

jiansowa replied:
done

jiansowa committed:
Define IMAGINIATION_NNA_SDK_ROOT with cmake configure_file instead of
hard coding in source code.

jiansowa committed:
Remove/edit some comments.

jiansowa committed:
Add imagination nna part in build_linux.sh. test=develop

jiansowa committed:
Use clang-foramt 3.8 insteady of 6.0.0, otherwise clang-format will
format memory_optimized_pass.cc again. test=develop

jiansowa committed:
Set LITE_WITH_IMAGINATION to OFF by defaults in build_linux.sh.

jiansowa committed:
1, Review the imgdnn_manager, make some changes and add assert at the
end of each IMGDNN call. test=develop

hong19860320 commented on README.md:
```diff
@@ -43,7 +43,6 @@ Paddle Lite提供了C++、Java、Python三种API，并且提供了相应API的
 - [iOS示例](https://paddle-lite.readthedocs.io/zh/latest/demo_guides/ios_app_demo.html)
 - [ARMLinux示例](https://paddle-lite.readthedocs.io/zh/latest/demo_guides/linux_arm_demo.html)
 - [X86示例](https://paddle-lite.readthedocs.io/zh/latest/demo_guides/x86.html)
-- [CUDA示例](https://paddle-lite.readthedocs.io/zh/latest/demo_guides/cuda.html)
```
为什么需要修改READM.md呢？

hong19860320 commented on docs/demo_guides/baidu_xpu.md:
```diff
@@ -16,69 +16,12 @@ Paddle Lite已支持百度XPU在x86和arm服务器（例如飞腾 FT-2000+/64）
 
 ### 已支持的Paddle模型
 
-- [ResNet50](https://paddlelite-demo.bj.bcebos.com/models/resnet50_fp32_224_fluid.tar.gz)
```
不要修改这个

hong19860320 commented on docs/demo_guides/cuda.md:
```diff
@@ -1,5 +1,7 @@
 # PaddleLite使用CUDA预测部署
 
+**注意**: Lite CUDA仅作为Nvidia GPU加速库，支持模型有限，如有需要请使用[PaddleInference](https://paddle-inference.readthedocs.io/en/latest)。
```
去掉这个文件的修改

jiansowa committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle-Lite into jiansowa/img_nna
Just before commit.

jiansowa committed:
Just to trigger a CI test. test=develop

jiansowa committed:
Remove the imagination nna sdk root from code, assume hwconf and
mapconf files located at directory where to run the program.

jiansowa committed:
Change the dir name of nna config files.

jiansowa committed:
Manually merge the missed part of imagination nna integration after pull
upstream develop. test=develop

jiansowa committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle-Lite into jiansowa/img_nna
test=develop

hong19860320 commented on lite/kernels/imagination_nna/bridges/conv_op.cc:
```diff
@@ -0,0 +1,278 @@
+  CHECK_EQ(output_dims[0], bs);
+  CHECK_EQ(output_dims[1], oc);
+  auto strides = op_info->GetAttr<std::vector<int>>("strides");
+  auto paddings = op_info->GetAttr<std::vector<int>>("paddings");
```
由于flatbuffer的加入，auto 需要改成std::vector<int>，paddings和dilations都需要修改

jiansowa replied:
Done.

jiansowa committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle-Lite into jiansowa/img_nna
[Imagination NNA] Explicitly set return type of GetAttr for std::vector
as template parameter. test=develop

hong19860320 commented on lite/backends/imagination_nna/imgdnn_manager.h:
```diff
@@ -0,0 +1,269 @@
+                                 int line,
+                                 const char *filename) {
+  if (cond) {
+    std::stringstream err_msg;
```
直接使用CHECK、CHECK_EQ等宏就可以了

jiansowa replied:
Done.

hong19860320 commented on lite/backends/imagination_nna/imgdnn_manager.h:
```diff
@@ -0,0 +1,269 @@
+  }
+}
+
+#define ASSERT(statement, msg) \
```
同上

jiansowa replied:
Done

jiansowa committed:
[Imagination NNA]
1, Use Assert of Paddle-Lite
2, Remove the redundant cmake config file.
test=develop

jiansowa committed:
[Imagination NNA] Remove the ASSERT macro not used. test=develop

hong19860320 reviewed (dismissed):
LGTM

Superjomn commented on lite/backends/imagination_nna/imgdnn_manager.h:
```diff
@@ -0,0 +1,253 @@
+  imgdnn_network_object net_obj_{nullptr};
+  std::vector<uint8_t *> coef_pool;
+
+  imgdnn_tensor convertQuantTensorType(imgdnn_tensor a_tensor,
```
Please follow google c++ code style, https://google.github.io/styleguide/cppguide.html#Function_Names

except the APIs from the third-parties.


Applicable to all the other functions in this PR.

jiansowa replied:
Done

Superjomn commented on lite/backends/imagination_nna/imgdnn_manager.h:
```diff
@@ -0,0 +1,253 @@
+  }
+
+  uint8_t *GetBufromPool(size_t size) {
+    uint8_t *buf = new uint8_t[size];
```
seems a std::vector<uint8_t> is better, no need to use a `new`?

avoid using a raw `new` or `delete`, wrap it with a smart pointer.

```c++
std::vector<uint8_t> pool;

uint8_t *GetBufromPool(size_t size) {
  pool.resize(pool.size()+size);
  return pool.data() + pool.size()-size;
}
```

jiansowa replied:
This is a good idea, and works in my small test program, but crash when integrate into paddle-lite tree. Using unique_ptr instead.

Superjomn commented on lite/core/context.h:
```diff
@@ -188,6 +189,20 @@ class Context<TargetType::kRKNPU> {
+  void CopySharedTo(ImaginationNNAContext* ctx) {}
+
+  // NNAContext& operator=(const NNAContext& ctx) {}
+  std::string name() const { return "ImaginationNNAContext"; }
```
const char*

jiansowa replied:
Need more information.

Superjomn commented on lite/kernels/imagination_nna/bridges/act_op.cc:
```diff
@@ -0,0 +1,73 @@
+  // Get input and output vars and op attributes
+  auto x_name = op_info->Input("X").front();
+  auto x_type = kernel->GetInputDeclType("X");
+  CHECK(x_type->precision() == PRECISION(kFloat));
```
can this work?  `CHECK_EQ(x_type->precision(), PRECISION(kFloat))`

jiansowa replied:
Actually, it's copied from reference code. I will use CHECK_EQ here.

jiansowa replied:
CHECK_EQ does not works, will cause compile error.

Superjomn commented on lite/kernels/imagination_nna/bridges/act_op.cc:
```diff
@@ -0,0 +1,73 @@
+  if (graph->Has(x_name)) {
+    x_node = graph->Get(x_name);
+  } else {
+    // x_node = graph->Add(x_name, *x);
```
clean

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/utility.h:
```diff
@@ -0,0 +1,50 @@
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "imgdnn.h"  // NOLINT
```
#include <imgdnn.h>

jiansowa replied:
Done.

Superjomn commented on lite/kernels/imagination_nna/bridges/softmax_op.cc:
```diff
@@ -0,0 +1,80 @@
+namespace imagination_nna {
+
+int SoftmaxConverter(void* ctx, OpLite* op, KernelBase* kernel) {
+  CHECK(ctx != nullptr);
```
`CHECK(ctx)` is enough here.

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/pool_op.cc:
```diff
@@ -0,0 +1,145 @@
+// limitations under the License.
+
+#include "lite/operators/pool_op.h"
+#include "imgdnn.h"  // NOLINT
```
`#include <imgdnn.h>`, applicable to all the header files from the third-parties.

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/fc_op.cc:
```diff
@@ -0,0 +1,171 @@
+      bias_node = graph->Get(bias_name);
+    } else {
+      auto bias = scope->FindTensor(bias_name);
+      // CHECK_EQ(bias->precision(), PRECISION(kFloat));
```
remove all the unneeded commented code.

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/fc_op.cc:
```diff
@@ -0,0 +1,171 @@
+  auto scope = op->scope();
+  VLOG(3) << "[NNA] Converting " + op_type + "...";
+
+  CHECK(op_info->HasAttr("enable_int8") &&
```
expand to two CHECKs:

```
CHECK(op_info->HasAttr("enable_int8");
CHECK(op_info->GetAttr<bool>("enable_int8"));
```

So it shows exactly which condition is broken when failed.

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/conv_op.cc:
```diff
@@ -0,0 +1,278 @@
+    }
+  }
+
+  unsigned int img_stride[2] = {(unsigned int)strides[0],
```
use C++ style `static_cast` to cast the type here.

jiansowa replied:
Done

Superjomn commented on lite/kernels/imagination_nna/bridges/conv_op.cc:
```diff
@@ -0,0 +1,278 @@
+  // 0: {oc}
+  // 1: {1, oc, oh, ow}
+  // 2: {n, oc, oh, ow}
+  std::shared_ptr<Node> bias_node = NULL;
```
use c++ style `nullptr` instead of `NULL`

No need to initialize a shared_ptr.

jiansowa replied:
Done

Superjomn commented on lite/kernels/CMakeLists.txt:
```diff
@@ -15,3 +15,4 @@ add_subdirectory(apu)
 add_subdirectory(bm)
 add_subdirectory(rknpu)
 add_subdirectory(huawei_ascend_npu)
+add_subdirectory(imagination_nna)
```
remove the binary file below

No binary file is allowed to merge to the codebase.

Superjomn commented on lite/backends/imagination_nna/imgdnn_manager.cc:
```diff
@@ -0,0 +1,366 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "imgdnn_manager.h"  // NOLINT
```
`#include <imgdnn_manager.h>`

jiansowa replied:
Done

jiansowa committed:
[Imagination NNA] Follow google c++ style. test=develop

Superjomn reviewed (approved):
LGTM

hong19860320 reviewed (approved):
LGTM
